### PR TITLE
python3Packages.llvmlite: make it statically link to llvm

### PR DESCRIPTION
**Pull request by wrvsrx**
## Description of changes

Llvmlite doc says that they link to llvm statically, but it's not. Check

- https://github.com/numba/llvmlite/issues/1027.
- https://github.com/numba/llvmlite/blob/ca123c3ae2a6f7db865661ae509862277ec5d692/docs/source/faqs.rst#why-static-linking-to-llvm

This behavior causes problems when python uses another version of llvm, for example, mesa llvmpipe. This PR force llvmlite statically link to llvm.

Besides, this PR also use llvm as buildInputs rather than nativeBuildInputs since it's a library that llvmlite use rather than a part of build system.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by NickCao** [comment]
I think we should make this an option, and cleanup the outdated patching.

```diff
--- a/pkgs/development/python-modules/llvmlite/default.nix
+++ b/pkgs/development/python-modules/llvmlite/default.nix
@@ -5,13 +5,11 @@
   buildPythonPackage,
   isPyPy,
   pythonAtLeast,
-
-  # build-system
   llvm,
+  libxml2,
   setuptools,
-
-  # tests
   python,
+  withStaticLLVM ? true,
 }:
 
 buildPythonPackage rec {
@@ -28,22 +26,16 @@ buildPythonPackage rec {
     hash = "sha256-5QBSRDb28Bui9IOhGofj+c7Rk7J5fNv5nPksEPY/O5o=";
   };
 
-  nativeBuildInputs = [
-    llvm
-    setuptools
-  ];
+  build-system = [ setuptools ];
 
-  # Disable static linking
-  # https://github.com/numba/llvmlite/issues/93
-  postPatch = ''
-    substituteInPlace ffi/Makefile.linux --replace "-static-libstdc++" ""
+  buildInputs = [ llvm ] ++ lib.optionals withStaticLLVM [ libxml2.dev ];
 
-    substituteInPlace llvmlite/tests/test_binding.py --replace "test_linux" "nope"
-  '';
+  env.LLVM_CONFIG = "${llvm.dev}/bin/llvm-config";
 
-  # Set directory containing llvm-config binary
-  preConfigure = ''
-    export LLVM_CONFIG=${llvm.dev}/bin/llvm-config
+  postPatch = lib.optionalString withStaticLLVM ''
+    substituteInPlace ffi/build.py \
+      --replace-fail 'llvm_config, "--system-libs --libs all"' \
+                     'llvm_config, "--system-libs --libs all --link-static"'
   '';
 
   checkPhase = ''

```

The output size is 2.3M when dynamic linking, but 110M when static linking.

**Comment by wrvsrx** [comment]
@NickCao Done.

**Comment by risicle** [comment]
Does successfully build for me on macos 12 x86_64 & no new reverse dependencies fail (with it enabled)

File: pkgs/development/python-modules/llvmlite/default.nix [hunk 1]
```diff
@@ -28,34 +30,31 @@ buildPythonPackage rec {
...
 
   # Set directory containing llvm-config binary
-  preConfigure = ''
-    export LLVM_CONFIG=${llvm.dev}/bin/llvm-config
-  '';
+  env.LLVM_CONFIG = "${llvm.dev}/bin/llvm-config";
 
   checkPhase = ''
     runHook preCheck
-    ${python.executable} runtests.py
```
Comment (NickCao) [inline pkgs/development/python-modules/llvmlite/default.nix:51]: Why this change?

File: pkgs/development/python-modules/llvmlite/default.nix [hunk 2]
```diff
@@ -28,34 +30,31 @@ buildPythonPackage rec {
...
+    python runtests.py
     runHook postCheck
   '';
 
   __impureHostDeps = lib.optionals stdenv.isDarwin [ "/usr/lib/libm.dylib" ];
 
-  passthru.llvm = llvm;
```
Comment (NickCao) [inline pkgs/development/python-modules/llvmlite/default.nix:57]: We should keep the passthru, since dynamic linking is still the default.

File: pkgs/development/python-modules/llvmlite/default.nix [hunk 3]
```diff
@@ -28,34 +30,31 @@ buildPythonPackage rec {
...
-  # Disable static linking
-  # https://github.com/numba/llvmlite/issues/93
-  postPatch = ''
-    substituteInPlace ffi/Makefile.linux --replace "-static-libstdc++" ""
+  buildInputs =
+    [ llvm ]
+    ++ lib.optionals withStaticLLVM [
+      zlib
+      libxml2.dev
+      ncurses
```
Comment (NickCao) [inline pkgs/development/python-modules/llvmlite/default.nix:40]: `ncurses` and `zlib` are already propagated by `llvm`

File: pkgs/development/python-modules/llvmlite/default.nix (hunk 1 above)
Comment (wrvsrx) [reply pkgs/development/python-modules/llvmlite/default.nix:51]: Because `buildPythonPackage` will add `python` to `PATH` automatically.

Comment (NickCao) [reply pkgs/development/python-modules/llvmlite/default.nix:51]: > Because `buildPythonPackage` will add `python` to `PATH` automatically.

Not the case for pypy and other python implementations?

```
nix-repl> pypy.executable
"pypy"
```

**Comment by NickCao** [comment]
Please resolve the merge conflict.

File: pkgs/development/python-modules/llvmlite/default.nix [hunk 4]
```diff
@@ -28,32 +30,27 @@ buildPythonPackage rec {
...
-  ];
+  build-system = [ setuptools ];
 
-  postPatch = ''
-    substituteInPlace llvmlite/tests/test_binding.py \
-      --replace-fail "test_linux" "nope"
+  buildInputs = [ llvm ] ++ lib.optionals withStaticLLVM [ libxml2.dev ];
+
+  postPatch = lib.optionalString withStaticLLVM ''
+    substituteInPlace ffi/build.py --replace "--system-libs --libs all" "--system-libs --libs --link-static all"
```
Comment (NickCao) [inline pkgs/development/python-modules/llvmlite/default.nix:38]:
```suggestion
    substituteInPlace ffi/build.py --replace-fail "--system-libs --libs all" "--system-libs --libs --link-static all"
```

**Review by NickCao — COMMENTED** [review]
Otherwise LGTM.

**Comment by wrvsrx** [comment]
@NickCao Done.

**Review by NickCao — APPROVED** [review]
Result of `nixpkgs-review pr 335416` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>10 packages marked as broken and skipped:</summary>
  <ul>
    <li>khoj</li>
    <li>khoj.dist</li>
    <li>piper-train</li>
    <li>piper-train.dist</li>
    <li>python311Packages.bambi</li>
    <li>python311Packages.bambi.dist</li>
    <li>python311Packages.clifford</li>
    <li>python311Packages.clifford.dist</li>
    <li>python312Packages.clifford</li>
    <li>python312Packages.clifford.dist</li>
  </ul>
</details>
<details>
  <summary>12 packages failed to build:</summary>
  <ul>
    <li>python311Packages.ffcv</li>
    <li>python311Packages.ffcv.dist</li>
    <li>python311Packages.vllm</li>
    <li>python311Packages.vllm.dist</li>
    <li>python311Packages.ydata-profiling</li>
    <li>python311Packages.ydata-profiling.dist</li>
    <li>python312Packages.gumath</li>
    <li>python312Packages.gumath.dist</li>
    <li>python312Packages.vllm</li>
    <li>python312Packages.vllm.dist</li>
    <li>python312Packages.ydata-profiling</li>
    <li>python312Packages.ydata-profiling.dist</li>
  </ul>
</details>
<details>
  <summary>233 packages built:</summary>
  <ul>
    <li>backgroundremover</li>
    <li>backgroundremover.dist</li>
    <li>molbar (python312Packages.molbar)</li>
    <li>molbar.dist (python312Packages.molbar.dist)</li>
    <li>openai-whisper (python312Packages.openai-whisper)</li>
    <li>openai-whisper.dist (python312Packages.openai-whisper.dist)</li>
    <li>pianotrans</li>
    <li>pianotrans.dist</li>
    <li>pitivi</li>
    <li>python311Packages.aeppl</li>
    <li>python311Packages.aeppl.dist</li>
    <li>python311Packages.aesara</li>
    <li>python311Packages.aesara.dist</li>
    <li>python311Packages.apricot-select</li>
    <li>python311Packages.apricot-select.dist</li>
    <li>python311Packages.array-api-compat</li>
    <li>python311Packages.array-api-compat.dist</li>
    <li>python311Packages.arviz</li>
    <li>python311Packages.arviz.dist</li>
    <li>python311Packages.asteroid-filterbanks</li>
    <li>python311Packages.asteroid-filterbanks.dist</li>
    <li>python311Packages.awkward</li>
    <li>python311Packages.awkward.dist</li>
    <li>python311Packages.coffea</li>
    <li>python311Packages.coffea.dist</li>
    <li>python311Packages.corner</li>
    <li>python311Packages.corner.dist</li>
    <li>python311Packages.correctionlib</li>
    <li>python311Packages.correctionlib.dist</li>
    <li>python311Packages.cppe</li>
    <li>python311Packages.cppe.dist</li>
    <li>python311Packages.dask-awkward</li>
    <li>python311Packages.dask-awkward.dist</li>
    <li>python311Packages.dask-glm</li>
    <li>python311Packages.dask-glm.dist</li>
    <li>python311Packages.dask-ml</li>
    <li>python311Packages.dask-ml.dist</li>
    <li>python311Packages.datashader</li>
    <li>python311Packages.datashader.dist</li>
    <li>python311Packages.dscribe</li>
    <li>python311Packages.dscribe.dist</li>
    <li>python311Packages.flowmc</li>
    <li>python311Packages.flowmc.dist</li>
    <li>python311Packages.galois</li>
    <li>python311Packages.galois.dist</li>
    <li>python311Packages.graphrag</li>
    <li>python311Packages.graphrag.dist</li>
    <li>python311Packages.graspologic</li>
    <li>python311Packages.graspologic.dist</li>
    <li>python311Packages.gumath</li>
    <li>python311Packages.gumath.dist</li>
    <li>python311Packages.hyppo</li>
    <li>python311Packages.hyppo.dist</li>
    <li>python311Packages.librosa</li>
    <li>python311Packages.librosa.dist</li>
    <li>python311Packages.llvmlite</li>
    <li>python311Packages.llvmlite.dist</li>
    <li>python311Packages.miasm</li>
    <li>python311Packages.miasm.dist</li>
    <li>python311Packages.molbar</li>
    <li>python311Packages.molbar.dist</li>
    <li>python311Packages.mplhep</li>
    <li>python311Packages.mplhep.dist</li>
    <li>python311Packages.numba</li>
    <li>python311Packages.numba-scipy</li>
    <li>python311Packages.numba-scipy.dist</li>
    <li>python311Packages.numba.dist</li>
    <li>python311Packages.numbaWithCuda</li>
    <li>python311Packages.numbaWithCuda.dist</li>
    <li>python311Packages.numpy-groupies</li>
    <li>python311Packages.numpy-groupies.dist</li>
    <li>python311Packages.openai-whisper</li>
    <li>python311Packages.openai-whisper.dist</li>
    <li>python311Packages.outlines</li>
    <li>python311Packages.outlines.dist</li>
    <li>python311Packages.piano-transcription-inference</li>
    <li>python311Packages.piano-transcription-inference.dist</li>
    <li>python311Packages.pyannote-audio</li>
    <li>python311Packages.pyannote-audio.dist</li>
    <li>python311Packages.pygmo</li>
    <li>python311Packages.pymatting</li>
    <li>python311Packages.pymatting.dist</li>
    <li>python311Packages.pymc</li>
    <li>python311Packages.pymc.dist</li>
    <li>python311Packages.pymoo</li>
    <li>python311Packages.pymoo.dist</li>
    <li>python311Packages.pynndescent</li>
    <li>python311Packages.pynndescent.dist</li>
    <li>python311Packages.pyscf</li>
    <li>python311Packages.pyscf.dist</li>
    <li>python311Packages.pytensor</li>
    <li>python311Packages.pytensor.dist</li>
    <li>python311Packages.quaternion</li>
    <li>python311Packages.quaternion.dist</li>
    <li>python311Packages.resampy</li>
    <li>python311Packages.resampy.dist</li>
    <li>python311Packages.scikit-tda</li>
    <li>python311Packages.scikit-tda.dist</li>
    <li>python311Packages.shap</li>
    <li>python311Packages.shap.dist</li>
    <li>python311Packages.simple-dftd3</li>
    <li>python311Packages.simple-dftd3.dist</li>
    <li>python311Packages.sparse</li>
    <li>python311Packages.sparse.dist</li>
    <li>python311Packages.stumpy</li>
    <li>python311Packages.stumpy.dist</li>
    <li>python311Packages.stytra</li>
    <li>python311Packages.stytra.dist</li>
    <li>python311Packages.timezonefinder</li>
    <li>python311Packages.timezonefinder.dist</li>
    <li>python311Packages.tinygrad</li>
    <li>python311Packages.tinygrad.dist</li>
    <li>python311Packages.torch-audiomentations</li>
    <li>python311Packages.torch-audiomentations.dist</li>
    <li>python311Packages.torchcrepe</li>
    <li>python311Packages.torchcrepe.dist</li>
    <li>python311Packages.torchlibrosa</li>
    <li>python311Packages.torchlibrosa.dist</li>
    <li>python311Packages.trackpy</li>
    <li>python311Packages.trackpy.dist</li>
    <li>python311Packages.tsfresh</li>
    <li>python311Packages.tsfresh.dist</li>
    <li>python311Packages.umap-learn</li>
    <li>python311Packages.umap-learn.dist</li>
    <li>python311Packages.uproot</li>
    <li>python311Packages.uproot.dist</li>
    <li>python311Packages.vector</li>
    <li>python311Packages.vector.dist</li>
    <li>python311Packages.xarray-einstats</li>
    <li>python311Packages.xarray-einstats.dist</li>
    <li>python312Packages.apricot-select</li>
    <li>python312Packages.apricot-select.dist</li>
    <li>python312Packages.array-api-compat</li>
    <li>python312Packages.array-api-compat.dist</li>
    <li>python312Packages.arviz</li>
    <li>python312Packages.arviz.dist</li>
    <li>python312Packages.asteroid-filterbanks</li>
    <li>python312Packages.asteroid-filterbanks.dist</li>
    <li>python312Packages.awkward</li>
    <li>python312Packages.awkward.dist</li>
    <li>python312Packages.coffea</li>
    <li>python312Packages.coffea.dist</li>
    <li>python312Packages.corner</li>
    <li>python312Packages.corner.dist</li>
    <li>python312Packages.correctionlib</li>
    <li>python312Packages.correctionlib.dist</li>
    <li>python312Packages.cppe</li>
    <li>python312Packages.cppe.dist</li>
    <li>python312Packages.dask-awkward</li>
    <li>python312Packages.dask-awkward.dist</li>
    <li>python312Packages.dask-glm</li>
    <li>python312Packages.dask-glm.dist</li>
    <li>python312Packages.dask-ml</li>
    <li>python312Packages.dask-ml.dist</li>
    <li>python312Packages.dscribe</li>
    <li>python312Packages.dscribe.dist</li>
    <li>python312Packages.galois</li>
    <li>python312Packages.galois.dist</li>
    <li>python312Packages.graphrag</li>
    <li>python312Packages.graphrag.dist</li>
    <li>python312Packages.graspologic</li>
    <li>python312Packages.graspologic.dist</li>
    <li>python312Packages.hyppo</li>
    <li>python312Packages.hyppo.dist</li>
    <li>python312Packages.librosa</li>
    <li>python312Packages.librosa.dist</li>
    <li>python312Packages.llvmlite</li>
    <li>python312Packages.llvmlite.dist</li>
    <li>python312Packages.miasm</li>
    <li>python312Packages.miasm.dist</li>
    <li>python312Packages.mplhep</li>
    <li>python312Packages.mplhep.dist</li>
    <li>python312Packages.numba</li>
    <li>python312Packages.numba-scipy</li>
    <li>python312Packages.numba-scipy.dist</li>
    <li>python312Packages.numba.dist</li>
    <li>python312Packages.numbaWithCuda</li>
    <li>python312Packages.numbaWithCuda.dist</li>
    <li>python312Packages.numpy-groupies</li>
    <li>python312Packages.numpy-groupies.dist</li>
    <li>python312Packages.outlines</li>
    <li>python312Packages.outlines.dist</li>
    <li>python312Packages.piano-transcription-inference</li>
    <li>python312Packages.piano-transcription-inference.dist</li>
    <li>python312Packages.pyannote-audio</li>
    <li>python312Packages.pyannote-audio.dist</li>
    <li>python312Packages.pygmo</li>
    <li>python312Packages.pymatting</li>
    <li>python312Packages.pymatting.dist</li>
    <li>python312Packages.pymoo</li>
    <li>python312Packages.pymoo.dist</li>
    <li>python312Packages.pynndescent</li>
    <li>python312Packages.pynndescent.dist</li>
    <li>python312Packages.pyscf</li>
    <li>python312Packages.pyscf.dist</li>
    <li>python312Packages.quaternion</li>
    <li>python312Packages.quaternion.dist</li>
    <li>python312Packages.resampy</li>
    <li>python312Packages.resampy.dist</li>
    <li>python312Packages.scikit-tda</li>
    <li>python312Packages.scikit-tda.dist</li>
    <li>python312Packages.shap</li>
    <li>python312Packages.shap.dist</li>
    <li>python312Packages.simple-dftd3</li>
    <li>python312Packages.simple-dftd3.dist</li>
    <li>python312Packages.sparse</li>
    <li>python312Packages.sparse.dist</li>
    <li>python312Packages.stumpy</li>
    <li>python312Packages.stumpy.dist</li>
    <li>python312Packages.stytra</li>
    <li>python312Packages.stytra.dist</li>
    <li>python312Packages.timezonefinder</li>
    <li>python312Packages.timezonefinder.dist</li>
    <li>python312Packages.tinygrad</li>
    <li>python312Packages.tinygrad.dist</li>
    <li>python312Packages.torch-audiomentations</li>
    <li>python312Packages.torch-audiomentations.dist</li>
    <li>python312Packages.torchcrepe</li>
    <li>python312Packages.torchcrepe.dist</li>
    <li>python312Packages.torchlibrosa</li>
    <li>python312Packages.torchlibrosa.dist</li>
    <li>python312Packages.trackpy</li>
    <li>python312Packages.trackpy.dist</li>
    <li>python312Packages.umap-learn</li>
    <li>python312Packages.umap-learn.dist</li>
    <li>python312Packages.uproot</li>
    <li>python312Packages.uproot.dist</li>
    <li>python312Packages.vector</li>
    <li>python312Packages.vector.dist</li>
    <li>python312Packages.xarray-einstats</li>
    <li>python312Packages.xarray-einstats.dist</li>
    <li>whisper-ctranslate2</li>
    <li>whisper-ctranslate2.dist</li>
  </ul>
</details>

All failures already present on master.
